File: JavaScript/1171-Remove Zero Sum Consecutive Nodes from Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var removeZeroSumSublists = function(head) {
    const dummy = new ListNode(0);
    dummy.next = head;
    const hmap = new Map()
    pSum = 0;
    hmap.set(pSum, dummy)
    
    while (head){
        pSum += head.val;
        if(hmap.has(pSum)){
            let tmp = pSum, start = hmap.get(pSum).next;
            while (start !== head){
                tmp += start.val;
                hmap.delete(tmp);
                start = start.next;
            }
            hmap.get(pSum).next = head.next;
        } else{
            hmap.set(pSum, head);
        }
        head = head.next;
    }
    return dummy.next;
};

File: 1171-Remove Zero Sum Consecutive Nodes from Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeZeroSumSublists(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        hmap = {0:dummy}
        _sum = 0
        
        while head:
            _sum += head.val
            if _sum in hmap:
# delete unnecessary keys
                tmp = _sum
                start = hmap[tmp].next
                while start!=head:
                    tmp += start.val
                    del hmap[tmp]
                    start = start.next
# connect
                hmap[_sum].next = head.next
            else:
                hmap[_sum] = head
            head = head.next
            
        return dummy.next